### PR TITLE
fix for #1136. fix server crashing on invalid certificate

### PR DESCRIPTION
**Pull request by hoancea**
fix for #1136. See steps listed on the issue in order to replicate the issue

### CODE REVIEW

**Comment by hoancea** [comment]
I signed the CLA twice already, yet it still appears as not signed on the issue, not sure why.

File: packages/node-opcua-server/source/opcua_server.ts [hunk 1]
```diff
@@ -1502,7 +1503,11 @@ export class OPCUAServer extends OPCUABaseServer {
         assert(userTokenSignature.signature instanceof Buffer, "expecting userTokenSignature to be a Buffer");
 
         // verify proof of possession by checking certificate signature & server nonce correctness
-        if (!verifySignature(serverCertificate, nonce, userTokenSignature, certificate, securityPolicy)) {
+        try {
```
Comment (erossignon) [inline packages/node-opcua-server/source/opcua_server.ts]: Can we fix verifySignature instance to make it return false instead of throwing ?

**Review by erossignon — CHANGES_REQUESTED** [review]

File: packages/node-opcua-server/source/opcua_server.ts (hunk 1 above)
Comment (hoancea) [reply packages/node-opcua-server/source/opcua_server.ts]: solved

File: packages/node-opcua-secure-channel/source/security_policy.ts [hunk 2]
```diff
@@ -31,11 +31,15 @@ import {
...
 } from "node-opcua-crypto";
 import { EncryptBufferFunc, SignBufferFunc } from "node-opcua-chunkmanager";
+import { make_warningLog } from "node-opcua-debug";
 
 // tslint:disable:no-empty
 function errorLog(...args: any[]) {
     /** */
 }
+const warningLog = make_warningLog(__filename);
+const doTraceChunk = process.env.NODEOPCUADEBUG && process.env.NODEOPCUADEBUG.indexOf("CHUNK") >= 0;
```
Comment (erossignon) [inline packages/node-opcua-secure-channel/source/security_policy.ts]: CHUNK is used for tracing low level packet chunking on the transport layer

File: packages/node-opcua-secure-channel/source/security_policy.ts [hunk 3]
```diff
@@ -584,7 +588,14 @@ export function verifySignature(
     assert(signature.signature instanceof Buffer);
     // This parameter is calculated by appending the clientNonce to the clientCertificate
     const dataToVerify = Buffer.concat([chain[0], receiverNonce]);
-    return cryptoFactory.asymmetricVerify(dataToVerify, signature.signature, senderCertificate);
+    try {
+        return cryptoFactory.asymmetricVerify(dataToVerify, signature.signature, senderCertificate);
+    } catch (e) {
+        if (doTraceChunk) {
```
Comment (erossignon) [inline packages/node-opcua-secure-channel/source/security_policy.ts]: Let's remove this if statemtn, a warning can be emitted regardless.

**Review by erossignon — CHANGES_REQUESTED** [review]

File: packages/node-opcua-secure-channel/source/security_policy.ts (hunk 2 above)
Comment (erossignon) [reply packages/node-opcua-secure-channel/source/security_policy.ts]: I would remove the use of doTraceChunk  alltogether

File: packages/node-opcua-secure-channel/source/security_policy.ts (hunk 3 above)
Comment (hoancea) [reply packages/node-opcua-secure-channel/source/security_policy.ts]: fixed in commit d413004

File: packages/node-opcua-secure-channel/source/security_policy.ts (hunk 2 above)
Comment (hoancea) [reply packages/node-opcua-secure-channel/source/security_policy.ts]: fixed in commit d413004

**Pull request merged**